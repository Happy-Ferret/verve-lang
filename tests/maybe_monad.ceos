type Maybe<T> {
  Just(T)
  Nothing()
}

interface Monad<M> {
  virtual return <T>(T) -> M<T>
  virtual bind <T, U>(M<T>, (T) -> M<U>) -> M<U>
}

implementation Monad<Maybe> {
  fn return(x) {
    Just(x)
  }

  fn bind(maybe, f) {
    match maybe {
      Nothing() => Nothing()
      Just(x) => f(x)
    }
  }
}

fn printMaybe(maybe: Maybe<Stringify>) -> Void {
  match maybe {
    Nothing() => print("Nothing")
    Just(x) => print(x)
  }
}

printMaybe(Just(5))
printMaybe(Nothing())

fn square(n: Int) -> Maybe<Int> {
  Just(n * n)
}

printMaybe(bind(bind(Just(3), square), square))
