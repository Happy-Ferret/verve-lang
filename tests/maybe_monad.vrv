type maybe<t> {
  Just(t)
  Nothing()
}

interface monad<m> {
  virtual return <t>(t) -> m<t>
  virtual bind <t, u>(m<t>, (t) -> m<u>) -> m<u>
}

implementation monad<maybe> {
  fn return(x) {
    Just(x)
  }

  fn bind(mb, f) {
    match mb {
      Nothing() => Nothing()
      Just(x) => f(x)
    }
  }
}

fn print_maybe(mb: maybe<printable>) -> void {
  match mb {
    Nothing() => print("Nothing")
    Just(x) => print(x)
  }
}

print_maybe(Just(5))
print_maybe(Nothing())

fn square(n: int) -> maybe<int> {
  Just(n * n)
}

print_maybe(bind(bind(Just(3), square), square))
