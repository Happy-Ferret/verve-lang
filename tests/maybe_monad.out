5
Nothing
81
